### PR TITLE
Use k8s.io/endpointslice library

### PR DESCRIPTION
**Pull request by carlory**
**Please provide a description of this PR:**

[KEP-3685](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3685-endpointslice-reconciler-to-staging/README.md ) moves the EndpointSlice reconciler to staging, so we can remove mannal copy-pasting of the code.

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/pod.go [hunk 1]
```diff
@@ -158,19 +131,19 @@ func (pc *PodCache) onEvent(_, pod *v1.Pod, ev model.Event) error {
 	key := config.NamespacedName(pod)
 	switch ev {
 	case model.EventAdd:
-		if shouldPodBeInEndpoints(pod) && IsPodReady(pod) {
+		if endpointsliceutil.ShouldPodBeInEndpoints(pod, false) && IsPodReady(pod) {
```
Comment (carlory) [inline pilot/pkg/serviceregistry/kube/controller/pod.go:134]: https://github.com/istio/istio/pull/42156 updated ShouldPodBeInEndpoints with latest k8s but the function signature is inconsistent with the upstream code https://github.com/kubernetes/kubernetes/blob/133be895e0de2994353703e082f92b6b946ac0e2/pkg/controller/util/endpoint/controller_utils.go#L141. Is it intentional or a bug? @hzxuzhonghu @ramaraochavali

The unit tests always pass when I switch `false` to `true`.

https://github.com/kubernetes/kubernetes/pull/90783 clarified behavior of publishNotReadyAddresses like the following:

```
(base) ➜  ~ kubectl explain svc.spec.publishNotReadyAddresses
KIND:       Service
VERSION:    v1

FIELD: publishNotReadyAddresses <boolean>


DESCRIPTION:
    publishNotReadyAddresses indicates that any agent which deals with endpoints
    for this Service should disregard any indications of ready/not-ready. The
    primary use case for setting this field is for a StatefulSet's Headless
    Service to propagate SRV DNS records for its Pods for the purpose of peer
    discovery. The Kubernetes controllers that generate Endpoints and
    EndpointSlice resources for Services interpret this to mean that all
    endpoints are considered "ready" even if the Pods themselves are not. Agents
    which consume only Kubernetes generated endpoints through the Endpoints or
    EndpointSlice resources can safely assume this behavior.
```


FYI: 
- https://github.com/kubernetes/kubernetes/blob/60c4c2b2521fb454ce69dee737e3eb91a25e0535/staging/src/k8s.io/endpointslice/reconciler.go#L171
- https://github.com/kubernetes/kubernetes/blob/60c4c2b2521fb454ce69dee737e3eb91a25e0535/pkg/controller/endpoint/endpoints_controller.go#L425

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/pod.go:134]: what do you mean by signature?

Comment (carlory) [reply pilot/pkg/serviceregistry/kube/controller/pod.go:134]: k/k: `func ShouldPodBeInEndpoints(pod *v1.Pod, includeTerminating bool) bool`
istio: `func shouldPodBeInEndpoints(pod *v1.Pod) bool `

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/pod.go:134]: It is intended, always ignore terminating pod

Comment (carlory) [reply pilot/pkg/serviceregistry/kube/controller/pod.go:134]: What will happen if we don't ignore it? the upstream https://github.com/kubernetes/kubernetes/blob/60c4c2b2521fb454ce69dee737e3eb91a25e0535/staging/src/k8s.io/endpointslice/reconciler.go#L171 are always considered "ready" even if the Pods themselves are not.

I'm not familiar with the Istio project, can you help me to clarify the above questions? Thanks!

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/pod.go:134]: This means when a pod is draining, we donot generate endpoint with it.

Comment (carlory) [reply pilot/pkg/serviceregistry/kube/controller/pod.go:134]: I see, thank you.

**Review by howardjohn — CHANGES_REQUESTED** [review]
> Things you should NOT do
> Expect compatibility. This repo is changing quickly in direct support of Kubernetes.

I am really not sure we should do this. "[A little copying is better than a little dependency.](https://go-proverbs.github.io/)"

**Pull request closed without merging**